[L1T] Restructure handling of Calorimeter input to Correlator Layer-1 + new TkElectrons in barrel (Phase-2)

cerminar opened the pull request:
#### PR description:

This PR rewrites the Correlator Layer-1 emulator code handling the Calorimeter input objects, enabling more realistic treatment of barrel and endcap data. 

In particular:
- drop the conversion of the input calorimeter objects to `PFCluster`, moving all needed manipulations (deserialization, calibrations and ID models to the Layer-1 emulator.
   - "decoded" calorimeter objects are now built starting from digitized words via unpackers according to most up-to-date interface documents
- write `PFCluster` collections directly from the CTL1 emulator AFTER the decoding storing bit-wise accurate information across the whole calorimenter. This allows acessing data in the final precision for performance studies and ML model training
   - modify the `PFCluster` data-format to store digitized objects in packed format. This allows retrieving quantized quantities in the ntuplization steps
- adds more realistic HGCal decoding (update to more recent interface document)
   - add emulation of new multiclass ID within the HGCal decoding step
- PFCandidate ref to source becomes polymorphic, allowing uniform interface across barrel and endcap 
- GCT EM and Hadronic emulated objects are now decoded and calibrated within the emulator
- updated emulator of the barrel "TDR" regionizer
- flexible packed binary format of calorimeter and track objects across barrel and edncap to optimize Firmware resource usage


e-gamma L1T:
- adds new composite ID model for the barrel cluster to track matching for building TkElectrons

#### PR validation:

We will soon post some comparison plots.

cerminar committed:
Initial prototype to integrate the HGC ID multiclass model

* EEmulticlass_tinymodel

jmitrevs committed:
Adding handling of gct emulated objects to CtL1 barrel emulator

cerminar committed:
Switch to emulated crystal clusters and store digitized version in PFClusters

cerminar committed:
Initial implementation of new barrel and endcap Composite ID model

* Restructure model configuration and handling

cerminar committed:
update geometry

cerminar committed:
fix decoded track collection for missing tracks

cerminar committed:
Initial implementation of HGC cluster encoding/decoding + multiclass ID evaluation

* Remove redundant conifer header

cerminar committed:
Add chi2 to Tk format for PF and start adding ID prob to calo clusters introducing custom packing/unpacking for barrel/endcap

cerminar committed:
Towards a more realistic implementation + some cleaning

cerminar committed:
Complete info into decoded PFClusters

cerminar committed:
Propagate multiclass weights and handles hoe and EM scenarios

cerminar committed:
Update to more recent geometry and input files

cerminar committed:
PFCluster used to store decoded calo info and TkObj specialized packing and unpacking

* Restructure PFCluster object to store the decoded calorimeter info across the whole detector plus some other fixes.
* Specialize TkObj packing for barrel and endcap

cerminar committed:
Fix eta phi bias in GCT clusters + some cosmetics

cerminar committed:
Switch off calibration for GCT EM clusters

cerminar committed:
Equip for endcap barrel dedicated packing and unpacking of calo objs

Overload most used patterns

cerminar committed:
Fix eta conversion and move Multiclass after energy manipulation

cerminar committed:
first attempt at integration of both HGC and GCT inputs

cerminar committed:
Complete emulation of HGC deserialization, ID and calibration

Initial implementation of emulation of pt corrections for HGC clusters

can now emulate both corrections or corrected pt

Update HGC model and add softmax

cerminar committed:
Complete emulation of compositeID_EB_v1 and reworking of code for Composite ID models

cerminar committed:
Start cleaning up comments

cerminar committed:
Keep supporting Idea GCT Had clusters (from PFClusters)

jmitrevs committed:
APx regionizer updates and pattern file writing (#20)

cerminar committed:
Fist step towards removing input conversion to PFCluster

cerminar committed:
Move emulator objects sources to L1Candidate

cerminar committed:
Introduce pt dependent WP for HGC Multiclass model

cerminar committed:
Fix GCT EM cluster deserialization

code format

Fix TkEm WP and GCT energy

cerminar committed:
Update HGC ID model to full endcap + cosmetics

cerminar committed:
cleanup deprecated code

jmitrevs committed:
throw pf clusters out of range for GCT clusters, fix converter emulator (#22)

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"17d047b9914270e4099e7afc98795a4251409e82":{"files":["L1Trigger/Phase2L1ParticleFlow/data/compositeID.json","L1Trigger/Phase2L1ParticleFlow/data/endcap.json","L1Trigger/Phase2L1ParticleFlow/data/my_prj.json","L1Trigger/Phase2L1ParticleFlow/interface/conifer.h","L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegalgo_ref.h","L1Trigger/Phase2L1ParticleFlow/python/l1TkEgAlgoEmulator_cfi.py","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/egamma/pftkegalgo_ref.cpp"],"squashed":false,"time":1746193562},"1c5e13d0692e431295ebbaed2c9e1f567635d5b4":{"files":["DataFormats/L1TParticleFlow/interface/PFCandidate.h","DataFormats/L1TParticleFlow/src/classes_def.xml","L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py"],"squashed":false,"time":1746193565},"1ef025cbd2c8ab9a39ecf1dd5be1ac12346f79ff":{"files":["L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py","L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_binaryFiles_cfg.py"],"squashed":false,"time":1746193711},"209fc15cfbb1fe4cfef87c8fb14615f013579c52":{"files":["DataFormats/L1TParticleFlow/interface/PFCluster.h","DataFormats/L1TParticleFlow/src/classes_def.xml","L1Trigger/Phase2L1ParticleFlow/data/multiclassID/multiclass_EE.json","L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h","L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h","L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc","L1Trigger/Phase2L1ParticleFlow/python/l1tPFClustersFromHGC3DClusters_cfi.py","L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc"],"squashed":false,"time":1746193559},"2563f03bed07ffb40e324ae8f87a14891d81908e":{"files":["L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegalgo_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/hgcalinput_ref.h","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py"],"squashed":false,"time":1746193868},"315fa8fc38fa80a0847b41ff3935570160c5c6a6":{"files":["L1Trigger/DemonstratorTools/src/utilities.cc","L1Trigger/Phase2L1ParticleFlow/interface/L1TCorrelatorLayer1PatternFileWriter.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/gcteminput_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/gcthadinput_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/regionizer/tdr_regionizer_elements_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/regionizer/tdr_regionizer_elements_ref.icc","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_patternWriters_cff.py","L1Trigger/Phase2L1ParticleFlow/src/L1TCorrelatorLayer1PatternFileWriter.cc","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/gcteminput_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/gcthadinput_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/regionizer/multififo_regionizer_ref.cpp","L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_binaryFiles_cfg.py"],"squashed":false,"time":1746193861},"36a3caea9804eb0ad4ec10ee2e53dc019b4c989e":{"files":["DataFormats/L1TParticleFlow/interface/layer1_emulator.h","DataFormats/L1TParticleFlow/src/layer1_emulator.cpp","L1Trigger/Phase2L1ParticleFlow/interface/ParametricResolution.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/gcteminput_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/gcthadinput_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/puppi/linpuppi_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/regionizer/tdr_regionizer_elements_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/regionizer/tdr_regionizer_elements_ref.icc","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/ParametricResolution.cc","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/gcteminput_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/gcthadinput_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/puppi/linpuppi_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/regionizer/tdr_regionizer_ref.cpp","L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_binaryFiles_cfg.py"],"squashed":false,"time":1746193855},"385496f65356f24b057faa9b19560aa9e7e1b7e2":{"files":["L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc"],"squashed":false,"time":1746193557},"48066d4e2ae26f574898cf7e3a807ded7a6a808e":{"files":["DataFormats/L1TParticleFlow/interface/layer1_emulator.h","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc"],"squashed":false,"time":1746193568},"48209d1253b15b6d10af8bd04bb9e0da48e87bac":{"files":["L1Trigger/Phase2L1ParticleFlow/data/multiclassID/multiclass_EE.json","L1Trigger/Phase2L1ParticleFlow/interface/corrector.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/hgcalinput_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/nn_activation.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/nnet_common.h","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/corrector.cc","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193863},"4cde6596670c9fc8e4b90bec608bca4eb6ff4f98":{"files":["L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h","L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc","L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc","L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_binaryFiles_cfg.py"],"squashed":false,"time":1746193872},"54106879f50515087587b0d4402c0a088a8a9ac5":{"files":["L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py"],"squashed":false,"time":1746193862},"773fa448801f8ce4c2a85ba0fa523fbeee9b5820":{"files":["L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegalgo_ref.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/hgcalinput_ref.h","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/egamma/pftkegalgo_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193574},"7780c0deac98d51a1176532ee048ccefc6c530eb":{"files":["L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/gcteminput_ref.cpp"],"squashed":false,"time":1746193869},"787aa06a7e84023eee537ccff0c5f2b33f0a7e15":{"files":["DataFormats/L1TParticleFlow/interface/datatypes.h","DataFormats/L1TParticleFlow/interface/layer1_objs.h","L1Trigger/Phase2L1ParticleFlow/data/egamma/compositeID_EB_v1.json","L1Trigger/Phase2L1ParticleFlow/data/egamma/compositeID_EE_v0.json","L1Trigger/Phase2L1ParticleFlow/interface/conifer.h","L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegalgo_ref.h","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1TkEgAlgoEmulator_cfi.py","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/egamma/pftkegalgo_ref.cpp"],"squashed":false,"time":1746193866},"7d23d87c86f1bb63b5adbd3b1ff290c01bbad8f5":{"files":["L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_binaryFiles_cfg.py"],"squashed":false,"time":1746193859},"8296adb259d281e6c6383ea16a01cf7059a3ae76":{"files":["DataFormats/L1TParticleFlow/interface/PFCluster.h","L1Trigger/Phase2L1ParticleFlow/data/egamma/compositeID_EB_v0.json","L1Trigger/Phase2L1ParticleFlow/data/egamma/compositeID_EE_v0.json","L1Trigger/Phase2L1ParticleFlow/data/egamma/compositeID_EE_v1.json","L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h","L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h","L1Trigger/Phase2L1ParticleFlow/interface/l1-converters/hgcalinput_ref.h","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1TkEgAlgoEmulator_cfi.py","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193567},"8758d0516300c3f5f608c6b8fc86de4ed950d56f":{"files":["DataFormats/L1TCorrelator/src/TkEm.cc","DataFormats/L1TParticleFlow/interface/PFCluster.h","L1Trigger/Phase2L1ParticleFlow/plugins/DeregionizerProducer.cc","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc","L1Trigger/Phase2L1ParticleFlow/src/egamma/pftkegalgo_ref.cpp"],"squashed":false,"time":1746193571},"8b5825a0f88e28d59563c8db6aa2f61edd716e90":{"files":["L1Trigger/Phase2L1ParticleFlow/data/multiclassID/multiclass_EE.json","L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193870},"8d70ba994123610a81ecee2c8220bed621f8cdc2":{"files":["L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193716},"a255e8283bac62dc23290be8c81d93c6652202c6":{"files":["L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193714},"a48a4fcd3016d03a379c110855b0669675e040ae":{"files":["DataFormats/L1TParticleFlow/interface/PFCluster.h","DataFormats/L1TParticleFlow/interface/layer1_objs.h","DataFormats/L1TParticleFlow/src/classes_def.xml","L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauEmulator.cc","L1Trigger/L1CaloTrigger/plugins/L1NNCaloTauProducer.cc","L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterEgID.h","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromHGC3DClusters.cc","L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromL1EGClusters.cc","L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterEgID.cc","L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc","L1Trigger/Phase2L1ParticleFlow/src/egamma/pfeginput_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/hgcalinputt_ref.cpp"],"squashed":false,"time":1746193645},"a95d6ec7121a599c2262a75478c1ea002cff4ff2":{"files":["DataFormats/L1TParticleFlow/interface/layer1_objs.h","L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc"],"squashed":false,"time":1746193572},"bb4ce02450667f79edd249d59c242cc293e82cb7":{"files":["L1Trigger/Phase2L1ParticleFlow/test/make_l1ct_binaryFiles_cfg.py"],"squashed":false,"time":1746193554},"c1d92103a8b9a78a270cbd68473a8799b17bc1b5":{"files":["DataFormats/L1TParticleFlow/interface/PFCluster.h","DataFormats/L1TParticleFlow/src/classes_def.xml","L1Trigger/Phase2L1ParticleFlow/plugins/PFClusterProducerFromL1EGClusters.cc","L1Trigger/Phase2L1ParticleFlow/python/l1tPFClustersFromL1EGClusters_cfi.py"],"squashed":false,"time":1746193561},"e9b59fb05437f1f16aa8b23d0a402332fa4b86e4":{"files":["DataFormats/L1TParticleFlow/interface/PFCluster.h","DataFormats/L1TParticleFlow/src/classes_def.xml","L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h","L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc"],"squashed":false,"time":1746193563},"f5738e869adc3ab56b7031acdd30bb9cfe9237d9":{"files":["DataFormats/L1TParticleFlow/interface/bit_encoding.h"],"squashed":false,"time":1746193715},"f5eae1010d25bc15ac89680fbaac6bbee904b511":{"files":["L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc","L1Trigger/Phase2L1ParticleFlow/src/egamma/pftkegalgo_ref.cpp"],"squashed":false,"time":1746193649},"fa71ef7a63698dff8ced69071407801f2621f6b0":{"files":["DataFormats/L1TParticleFlow/interface/datatypes.h","DataFormats/L1TParticleFlow/interface/layer1_emulator.h","DataFormats/L1TParticleFlow/interface/layer1_objs.h","L1Trigger/Phase2L1ParticleFlow/src/L1TCorrelatorLayer1PatternFileWriter.cc","L1Trigger/Phase2L1ParticleFlow/src/l1-converters/tkinput_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/pf/pfalgo2hgc_ref.cpp","L1Trigger/Phase2L1ParticleFlow/src/pf/pfalgo3_ref.cpp"],"squashed":false,"time":1746193570}},"emoji":{},"last_seen_sha":"4cde6596670c9fc8e4b90bec608bca4eb6ff4f98","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44675

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44675/cms-checkout-topic.log

cerminar committed:
Introduce tight WP for HGC ID

cerminar committed:
New WPs for endcap Composite TkEle

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44696

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44696/cms-checkout-topic.log

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44698

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 205ac24ac4d957b56f0ba299d3879eed83832160
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 45a5419a724be5271d82b12efe111e88c376dcbb, 5ba67e66c7bab18cbca9838096d509a9eeb95256
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/datatypes.h  modified in PR(s): #47792
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #47792
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44698/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44698/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44698/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44698/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cerminar committed:
code checks

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44699

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 205ac24ac4d957b56f0ba299d3879eed83832160
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 45a5419a724be5271d82b12efe111e88c376dcbb, 5ba67e66c7bab18cbca9838096d509a9eeb95256
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/datatypes.h  modified in PR(s): #47792
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #47792
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

cmsbuild commented:
A new Pull Request was created by  @cerminar for master.

It involves the following packages:

- DataFormats/L1TCorrelator (**upgrade, l1**)
- DataFormats/L1TParticleFlow (**upgrade, l1**)
- L1Trigger/DemonstratorTools (**l1**)
- L1Trigger/L1CaloTrigger (**upgrade, l1**)
- L1Trigger/Phase2L1ParticleFlow (**upgrade, l1**)


@BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @missirol, @mmusich, @rovere this is something you requested to watch as well.
@antoniovilela, @mandrenguyen, @rappoccio, @sextonkennedy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 868KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45857/summary.html
**COMMIT**: f818cee83ae850b9327bb2b976181e14f15b72f5
**CMSSW**: CMSSW_15_1_X_2025-05-05-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48002/45857/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

quinnanm commented:
test parameters:
  - workflow = 29634.78
  - relvals_opt = --what upgrade

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 868KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45858/summary.html
**COMMIT**: f818cee83ae850b9327bb2b976181e14f15b72f5
**CMSSW**: CMSSW_15_1_X_2025-05-05-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48002/45858/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

jmitrevs committed:
remove GCT duplicates when using Ideal regionizer (#23)

cerminar committed:
Fix clang warnings

cerminar committed:
Fix bit ordering in HGC hwqual

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44730

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 205ac24ac4d957b56f0ba299d3879eed83832160
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 45a5419a724be5271d82b12efe111e88c376dcbb, 5ba67e66c7bab18cbca9838096d509a9eeb95256
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/datatypes.h  modified in PR(s): #47792
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #47792
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

cmsbuild commented:
Pull request #48002 was updated. @BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please check and sign again.

BenjaminRS commented:
please test

cmsbuild commented:
-1

**Failed Tests**: RelVals RelVals-INPUT 
**Size**: This PR adds an extra 28KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/summary.html
**COMMIT**: b18f1e21a1e77aa00e51d0696bdfb6ce0768631e
**CMSSW**: CMSSW_15_1_X_2025-05-06-2300/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48002/45902/install.sh` to create a dev area with all the needed externals and cmssw changes.


## RelVals

- [29634.75](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29634.75_TTbar_14TeV+Run4D110_HLT75e33Timing)
```
----- Begin Fatal Exception 07-May-2025 11:47:18 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'HLTriggerFinalPath'
   [2] Prefetching for module TriggerSummaryProducerAOD/'hltTriggerSummaryAOD'
   [3] Prefetching for module L1HPSPFTauProducer/'l1tHPSPFTauProducer'
   [4] Prefetching for module L1TPFCandMultiMerger/'l1tLayer1'
   [5] Calling method for module L1TCorrelatorLayer1Producer/'l1tLayer1Barrel'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<std::vector<l1tp2::GCTEmDigiCluster> >
Looking for module label: l1tPhase2GCTBarrelToCorrelatorLayer1Emulator
Looking for productInstanceName: GCTEmDigiClusters

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```
- [29834.999](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29834.999_TTbar_14TeV+Run4D110PU_PMXS1S2PR)
```
----- Begin Fatal Exception 07-May-2025 11:51:59 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'HLTriggerFinalPath'
   [2] Prefetching for module TriggerSummaryProducerAOD/'hltTriggerSummaryAOD'
   [3] Prefetching for module L1HPSPFTauProducer/'l1tHPSPFTauProducer'
   [4] Prefetching for module L1TPFCandMultiMerger/'l1tLayer1'
   [5] Calling method for module L1TCorrelatorLayer1Producer/'l1tLayer1Barrel'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<std::vector<l1tp2::GCTEmDigiCluster> >
Looking for module label: l1tPhase2GCTBarrelToCorrelatorLayer1Emulator
Looking for productInstanceName: GCTEmDigiClusters

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```
- [24834.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/24834.911_TTbar_14TeV+Run4D98_DD4hep)
```
----- Begin Fatal Exception 07-May-2025 11:55:00 CEST-----------------------
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 1 lumi: 1 event: 1 stream: 0
   [1] Running path 'HLTriggerFinalPath'
   [2] Prefetching for module TriggerSummaryProducerAOD/'hltTriggerSummaryAOD'
   [3] Prefetching for module L1HPSPFTauProducer/'l1tHPSPFTauProducer'
   [4] Prefetching for module L1TPFCandMultiMerger/'l1tLayer1'
   [5] Calling method for module L1TCorrelatorLayer1Producer/'l1tLayer1Barrel'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: std::vector<std::vector<l1tp2::GCTEmDigiCluster> >
Looking for module label: l1tPhase2GCTBarrelToCorrelatorLayer1Emulator
Looking for productInstanceName: GCTEmDigiClusters

   Additional Info:
      [a] If you wish to continue processing events after a ProductNotFound exception,
add "TryToContinue = cms.untracked.vstring('ProductNotFound')" to the "options" PSet in the configuration.

----- End Fatal Exception -------------------------------------------------
```
<details>
<summary>Expand to see more relval errors ...</summary>

- [29700.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29700.0_CloseByPGun_CE_H_Coarse_Scint+Run4D110)
- [29634.78](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29634.78_TTbar_14TeV+Run4D110_L1Complete)
- [29696.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29696.0_CloseByPGun_CE_E_Front_120um+Run4D110)
- [29634.911](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29634.911_TTbar_14TeV+Run4D110_DD4hep)
- [29634.0](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/45902/runTheMatrix-results/29634.0_TTbar_14TeV+Run4D110)
</details>



## RelVals-INPUT

- 134.0```DAS Error```

cerminar committed:
fix Phase-2 L1T configuration for GCT emulated objects

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44736

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 205ac24ac4d957b56f0ba299d3879eed83832160
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 2faf40ac662e7008386c0c59c5ce40f3b49e0c1e
    - Modified: bfd76ba02fa0f1aa6f386fca472dbe319ce11f0d, 45a5419a724be5271d82b12efe111e88c376dcbb, 5ba67e66c7bab18cbca9838096d509a9eeb95256
    - Deleted: 41b5f24f20a3768325fcbdd9917f01c1ab6f425d
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: bba28e5a7ef18fffad4c3664aca4377f7ff546d3
    - Deleted: 205ac24ac4d957b56f0ba299d3879eed83832160

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1TParticleFlow/interface/datatypes.h  modified in PR(s): #47792
  - File  DataFormats/L1TParticleFlow/src/classes_def.xml  modified in PR(s): #47792
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

cmsbuild commented:
Pull request #48002 was updated. @BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please check and sign again.

BenjaminRS commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44769

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Modified: da6d5e809eca1882236bdfd52efa10789be950bb, b970a9093f72b075c69d11a9aa587f6591e369e9
    - Deleted: 72b2a8dd526c4703bd396553ed9f253244d4bf1e
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Modified: da6d5e809eca1882236bdfd52efa10789be950bb, 2f2e58c3cde32c477ebe4b96f7e1ad630d7c7147, d8ed941273f3c152b9758649688a5574af4f21a4
    - Deleted: 72b2a8dd526c4703bd396553ed9f253244d4bf1e
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

cmsbuild commented:
Pull request #48002 was updated. @BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44799

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Modified: da6d5e809eca1882236bdfd52efa10789be950bb, b970a9093f72b075c69d11a9aa587f6591e369e9
    - Deleted: 72b2a8dd526c4703bd396553ed9f253244d4bf1e
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Modified: da6d5e809eca1882236bdfd52efa10789be950bb, 2f2e58c3cde32c477ebe4b96f7e1ad630d7c7147, d8ed941273f3c152b9758649688a5574af4f21a4
    - Deleted: 72b2a8dd526c4703bd396553ed9f253244d4bf1e
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

cmsbuild commented:
Pull request #48002 was updated. @BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please check and sign again.

cerminar commented:
Following up on the menu MET (thanks @RobertJWard ) rates at the DPG we discovered a missing commit in the rebase to 151X. The issue is now addressed.
Validation plots are now available  [here](https://cerminar.web.cern.ch/cerminar/plots/152X_validation_140Xv0D0/)  comparing: 
AR 2024 on 131X samples
151X vanilla on 140X samples
151X+PR on 140X samples
In particular:
MET rates: https://cerminar.web.cern.ch/cerminar/plots/152X_validation_140Xv0D0/jetmet_rate 
e/g rates: https://cerminar.web.cern.ch/cerminar/plots/152X_validation_140Xv0D0/ctl2_eg_rate 
e/g efficiencies: https://cerminar.web.cern.ch/cerminar/plots/152X_validation_140Xv0D0/ctl2_eg_eff 
e/g scale and resolution: https://cerminar.web.cern.ch/cerminar/plots/152X_validation_140Xv0D0/ctl2_eg_reso 

*Short summary*: no showstoppers. 

*More detailed summary*: big increase in efficiency for most of e/g objects at the price of some increase in rate (some of the menu WPs will need to be changed). Some increase in rate of PUPPI-MET. Seems related to a change in scale of the barrel. This change is related to the new GCT emulated egamma clusters and it was already seen in previous studies:
https://cerminar.web.cern.ch/cerminar/plots/fp131Xv3-CMSSW14.0.X-GCTEmu/ctl2_eg_reso/

BenjaminRS commented:
please test

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/46141/summary.html
**COMMIT**: b37188028045afdb82c05d5b74be6463e6386c52
**CMSSW**: CMSSW_15_1_X_2025-05-14-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48002/46141/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 6 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-14-1100+693570/68494/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 64 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-14-1100+693570/68494/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 4167455
 - DQMHistoTests: [Total failures: 751](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-14-1100+693570/68494/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4166684
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 220 log files, 188 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-05-14-1100+693570/68494/triggerResults): found differences in 8 / 49 workflows

quinnanm reviewed (commented):
Thank you for this PR! I just had a few minor comments/questions.

jmitrevs committed:
Improve TDR regionizer modeling of empty data in packets; fix issues in the review (#24)

Address review comments + fixes for TDR regionizer (not used in default CMSSW configuration)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/44966

- Found files with invalid states:
  - L1Trigger/Phase2L1ParticleFlow/data/compositeID.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/interface/HGC3DClusterID.h:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Modified: da6d5e809eca1882236bdfd52efa10789be950bb, b970a9093f72b075c69d11a9aa587f6591e369e9
    - Deleted: 72b2a8dd526c4703bd396553ed9f253244d4bf1e
  - L1Trigger/Phase2L1ParticleFlow/interface/conifer_cpp.h:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/data/endcap.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9
  - L1Trigger/Phase2L1ParticleFlow/src/HGC3DClusterID.cc:
    - Added: 18b68be31981ed3197b645363298581e5ba90409
    - Modified: da6d5e809eca1882236bdfd52efa10789be950bb, 2f2e58c3cde32c477ebe4b96f7e1ad630d7c7147, d8ed941273f3c152b9758649688a5574af4f21a4
    - Deleted: 72b2a8dd526c4703bd396553ed9f253244d4bf1e
  - L1Trigger/Phase2L1ParticleFlow/data/my_prj.json:
    - Added: 00f4a2930888a998f68dbcd46d170904385a062d
    - Deleted: b970a9093f72b075c69d11a9aa587f6591e369e9

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Phase2L1ParticleFlow/interface/puppi/linpuppi_ref.h  modified in PR(s): #48179
  - File  L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py  modified in PR(s): #48179
  - File  L1Trigger/Phase2L1ParticleFlow/src/puppi/linpuppi_ref.cpp  modified in PR(s): #48179
  - File  L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py  modified in PR(s): #47067

cerminar commented on L1Trigger/Phase2L1ParticleFlow/interface/egamma/pftkegalgo_ref.h:
```diff
@@ -34,7 +36,17 @@ namespace l1ct {
-    bool doCompositeTkEle;
+
+    enum Algo {
+      undefined = -1,
+      elliptic = 0,
+      compositeEE_v0 = 1,
+      compositeEB_v0 = 2,
+      compositeEE_v1 = 3,
+      compositeEB_v1 = 4
```
These numbers have no particular meaning, I am just assigning values to the `enum` to make sure that future changes will not break the correspondance with the `int` value set in the CMSSW cfg.

cerminar committed:
Tune TkEle WP + review comments

quinnanm commented:
please test

BenjaminRS commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:
```diff
@@ -478,36 +531,62 @@ void L1TCorrelatorLayer1Producer::produce(edm::Event &iEvent, const edm::EventSe
+    auto caloHandle = iEvent.getHandle(hadHGCalCands_);
+    if (caloHandle.isValid()) {
+      const auto &calos = *caloHandle;
+      for (unsigned int ic = 0, nc = calos.size(); ic < nc; ++ic) {
```
Why define `nc` here? Also here: https://github.com/cms-sw/cmssw/blob/d29d4100ada0be4b96ae53eb1085eae4bb1ae7cc/L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc#L573

cerminar replied:
This is a relatively common idiom to cast `size_t` explicitly. Anyway, I see no problem with caching the size here, no?

BenjaminRS replied:
I was just curious why it wasn't just:
`for (unsigned int ic = 0; ic < calos.size(); ++ic) {`
or even something like:
`for (auto ic : calos) {`
?

BenjaminRS commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:
```diff
@@ -636,6 +715,137 @@ void L1TCorrelatorLayer1Producer::produce(edm::Event &iEvent, const edm::EventSe
+  // FIXME: we keep subtracting an arbitrary number different from the HLGCal FW one
+  ap_ufixed<12, 11, AP_RND_CONV, AP_SAT> w_meanz = fabs(c.zBarycenter()) - 320;  // LSB = 0.5cm
+
+  ap_uint<6> w_showerlenght = c.showerLength();
```
`w_showerlength` and https://github.com/cms-sw/cmssw/blob/d29d4100ada0be4b96ae53eb1085eae4bb1ae7cc/L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc#L769

BenjaminRS commented on L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:
```diff
@@ -636,6 +715,137 @@ void L1TCorrelatorLayer1Producer::produce(edm::Event &iEvent, const edm::EventSe
+  ap_uint<6> w_showerlenght = c.showerLength();
+  ap_uint<7> w_sigmazz = round(c.sigmaZZ() / SIGMAZZ_LSB);
+  ap_uint<7> w_sigmaphiphi = round(c.sigmaPhiPhiTot() / SIGMAPHIPHI_LSB);
+  ap_uint<6> w_coreshowerlenght = c.coreShowerLength();
```
`w_coreshowerlength` and https://github.com/cms-sw/cmssw/blob/d29d4100ada0be4b96ae53eb1085eae4bb1ae7cc/L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc#L772

BenjaminRS reviewed (commented):
Hi Gianluca, thanks for the PR. Along with moving the jsons to externals here are some more comments inline:

cerminar committed:
tune HGC WPs to recover jet eff. and lower eg. sta rate

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/45153

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/45153/cms-checkout-topic.log

cmsbuild commented:
Pull request #48002 was updated. @BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please check and sign again.

BenjaminRS commented:
please test
(will run a test command now and wait for the final changes wrt the review, expected within the next 24hrs)

cmsbuild commented:
-1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/46691/summary.html
**COMMIT**: ba78e43ce8ccbff852d26549bf68229b898212a6
**CMSSW**: CMSSW_15_1_X_2025-06-12-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/48002/46691/install.sh` to create a dev area with all the needed externals and cmssw changes.

This pull request cannot be automatically merged, could you please rebase it?
You can see the log for git cms-merge-topic here: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/46691/git-merge-result

cerminar committed:
Address review comments

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/45171

ERROR: Unable to merge PR.

See log https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48002/45171/cms-checkout-topic.log

jmitrevs committed:
add comment on tdr regionizer constants

cerminar committed:
default constructor as per review

cerminar committed:
Move conifer json files to cms-data repo

cerminar commented:
The data files have been moved to `cms-data`: PR https://github.com/cms-data/L1Trigger-Phase2L1ParticleFlow/pull/8

cmsbuild commented:
Pull request #48002 was updated. @BenjaminRS, @Moanwar, @cmsbuild, @quinnanm, @srimanob, @subirsarkar can you please check and sign again.

BenjaminRS commented:
test parameters:
  - workflow = 29634.78
  - pull_request = cms-data/L1Trigger-Phase2L1ParticleFlow#8
  - relvals_opt = --what upgrade

BenjaminRS commented:
please test

BenjaminRS reviewed (approved):
OK thanks for making the changes

quinnanm commented:
+l1

Moanwar commented on L1Trigger/Phase2L1ParticleFlow/test/make_l1ctLayer1_dumpFiles_fromRAW_cfg.py:
```diff
@@ -15,11 +15,11 @@
-process.load('Configuration.Geometry.GeometryExtendedRun4D49_cff')
+process.load('Configuration.Geometry.GeometryExtended2026D95Reco_cff')
+process.load('Configuration.Geometry.GeometryExtended2026D95_cff')
 process.load('Configuration.StandardSequences.MagneticField_cff')
```
Just to checking are we still using the GeometryExtended2026 not Run4D110 or Run4D121 ?

mmusich replied:
I *guess* the issue is that `GeometryExtended2026D95_cff` does not [exist](https://github.com/search?q=repo%3Acms-sw%2Fcmssw%20GeometryExtended2026D95_cff&type=code) whereas `GeometryExtendedRun4D95_cff` [does]( https://github.com/search?q=repo%3Acms-sw%2Fcmssw%20GeometryExtendedRun4D95_cff&type=code) :)

Moanwar replied:
Yes, exactly — that’s what I wanted to check. Regardless of which geometry scenario that will be used — whether D95, D110, or another — it’s no longer referred to as '2026'. The files have been updated to something like 'Run4...' instead.

cerminar committed:
update geometry

BenjaminRS commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/46766/summary.html
**COMMIT**: 67e188a7674235f352b6d39d790a9c1ddd8f1b0e
**CMSSW**: CMSSW_15_1_X_2025-06-16-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48002/46766/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestDQMGUIUpload had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 25 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68991/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 64 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68991/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 4181335
 - DQMHistoTests: [Total failures: 1609](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68991/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4179706
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 220 log files, 188 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68991/triggerResults): found differences in 8 / 49 workflows

BenjaminRS commented:
@cms-sw/externals-l2 - can we ignore this Unit test failure as it seems to not be connected to this PR (which had [previously passed tests](https://github.com/cms-sw/cmssw/pull/48002#issuecomment-2970854633) before a [2 line configuration change to the recent Geometry files](https://github.com/cms-sw/cmssw/pull/48002/commits/67e188a7674235f352b6d39d790a9c1ddd8f1b0e))?

BenjaminRS commented:
please test
(Launching again overnight)

cmsbuild commented:
+1

**Size**: This PR adds an extra 20KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-693570/46778/summary.html
**COMMIT**: 67e188a7674235f352b6d39d790a9c1ddd8f1b0e
**CMSSW**: CMSSW_15_1_X_2025-06-16-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48002/46778/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [added 26 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68997/validateJR/logRootQA-events.log) to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 60 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68997/validateJR.html)
 - DQMHistoTests: Total files compared: 51
 - DQMHistoTests: Total histograms compared: 4181335
 - DQMHistoTests: [Total failures: 1606](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68997/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4179709
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 50 files compared)
 - Checked 220 log files, 188 edm output root files, 51 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-06-16-1100+693570/68997/triggerResults): found differences in 8 / 49 workflows

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @rappoccio, @antoniovilela, @sextonkennedy, @mandrenguyen (and backports should be raised in the release meeting by the corresponding L2)
**Notice** This PR was tested with additional Pull Request(s), please also merge them if necessary: cms-data/L1Trigger-Phase2L1ParticleFlow#8

mandrenguyen commented:
+1
